File: v1/src/main/java/com/api/v1/customers/services/CustomerRetrievalServiceImpl.java
package com.api.v1.customers.services;

import com.api.v1.customers.domain.CustomerRepository;
import com.api.v1.customers.dtos.CustomerResponseDto;
import com.api.v1.customers.utils.CustomerResponseMapperUtil;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

@Service
class CustomerRetrievalServiceImpl implements CustomerRetrievalService {

    @Autowired
    private CustomerRepository customerRepository;

    @Override
    public Flux<CustomerResponseDto> findAll() {
        return customerRepository
                .findAll()
                .flatMap(customer -> Mono.just(CustomerResponseMapperUtil.mapToDto(customer)));
    }

}

File: kotlin/v3/src/main/kotlin/com/api/v3/customers/services/CustomerRetrievalServiceImpl.kt
package com.api.v3.customers.services

import com.api.v3.customers.domain.CustomerRepository
import com.api.v3.customers.dtos.CustomerResponseDto
import com.api.v3.customers.utils.CustomerFinderUtil
import com.api.v3.customers.utils.CustomerResponseMapper
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.withContext
import org.springframework.stereotype.Service

@Service
internal class CustomerRetrievalServiceImpl(
    private val customerFinderUtil: CustomerFinderUtil,
    private val customerRepository: CustomerRepository
): CustomerRetrievalService {

    override suspend fun findBySsn(ssn: String): CustomerResponseDto {
        return withContext(Dispatchers.IO) {
            val existingCustomer = customerFinderUtil.find(ssn)
            CustomerResponseMapper.map(existingCustomer)
        }
    }

    override suspend fun findAll(): Flow<CustomerResponseDto> {
        return withContext(Dispatchers.IO) {
            customerRepository
                .findAll()
                .map { e -> CustomerResponseMapper.map(e) }
        }
    }

}